colossalai/inference/batch_bucket.py

```diff
@@ -0,0 +1,459 @@
+            and block_tables (List[torch.Tensor]): block tables of the target sequences indicating corresponding blocks
+        """
+        seqs = []
+        block_tables = []
+        for request_id in request_ids:
+            seq, block_table = self.pop_seq_update_batch(request_id, free_block_table_fn)
+            if seq is not None:
+                seqs.append(seq)
+            if block_table is not None:
+                block_tables.append(block_table)
+        return seqs, block_tables
+
+    def pop_n_seqs(
+        self, n: int, free_block_table_fn: Callable[[torch.Tensor], None] = None
+    ) -> Tuple[List[Sequence], List[torch.Tensor]]:
+        """Pop the first n sequences in the batch (FIFO).
+        If n is greater than the current batch szie, pop all the sequences in the batch.
+
+        Args:
+            n (int): The number of sequences to pop out
+            free_block_table_fn (Callable): The function to free the block table of a single sequence
+        Returns:
+            A tuple of: seqs (List[Sequence]): The target sequences,
+            and block_tables (List[torch.Tensor]): block tables of the target sequences indicating corresponding blocks
+        """
+        # NOTE Prevent calling this method multiple times in a single step
+        seqs = []
+        block_tables = []
+        n = min(n, self.current_batch_size)
+        seq_ids = list(self._sequences_dict.keys())[:n]
+        for seq_id in seq_ids:
+            seq = self._sequences_dict.pop(seq_id)
+            seq_b_idx = self._sequences_indexes.pop(seq_id)
+            if free_block_table_fn:
+                free_block_table_fn(self.block_tables[seq_b_idx])
+            else:
+                block_tables.append(self.block_tables[seq_b_idx].detach().clone())
+            seqs.append(seq)
+        if not self.is_compact:
+            self._make_compact()
+        return seqs, block_tables
+
+    def pop_finished(
+        self, free_block_table_fn: Callable[[torch.Tensor], None] = None
+    ) -> Tuple[List[Sequence], List[torch.Tensor]]:
+        """Pop finished sequences in the batch and a list of block tables of the finished sequences,
+        if free_block_table_fn is not provided.
+
+        Args:
+            free_block_table_fn (Callable): The function to free the block table of a single sequence
+        Returns:
+            A tuple of: finished_seqs (List[Sequence]): The finished sequences,
+            and finished_block_tables (List[torch.Tensor]): block tables of the finished sequences.
+        """
+        finished_seqs = []
+        finished_block_tables = []
+        for seq in self._sequences_dict.values():
+            if seq.check_finish():
+                finished_seqs.append(seq)
+
+        # Use `pop_seq_update_batch`` to update the batch status for just a few of finished seqs,
+        # otherwise, pop seqs directly and then call `_make_compact` to compress the batch
+        if len(finished_seqs) < 5:
```

Comment:
Why do we use 5 here?

Reply:
A constant to be set here. When we have relatively less finished seqs (namely just several seqs), use the first method to pop and replace; when we have more finished seqs, use the second way to pop.

Reply:
How was the number 5 chosen here?

Reply:
Randomly select a small value. Precise evaluations to be done but for now 3 or 6 or whatever small value works fine.

Reply:
Use the first method for now. Comments added.